### ISSUE TITLE
themes not working as documented

### ISSUE THREAD

**Issue by tbnorth** [leo-editor/leo-editor#350]
Using HOME=/tmp python launchLeo.py to operate in a clean environment, pasting a theme node from leoSettings.leo to my local settings doesn't enable the theme.  Not surprising as they don't contain @data qt-gui-plugin-style-sheet nodes.  Original theme machinery has a button to build that node from the Leonine tree based representation of the theme.

**Comment by edreamleo** [comment]
@tbnorth I'm confused by all this.  There *is* an `@data qt-gui-plugin-style-sheet` node in leoSettings.leo.  Are you saying that each theme must have its own `@data qt-gui-plugin-style-sheet` node?

If so, wouldn't it best to provide such a node for each theme in the `@ignore Themes` tree? Why don't we just do that?

It looks like neither Leo 5.3 nor 5.4 had per-theme @data qt-gui-plugin-style-sheet nodes, so this issue has been around a long, long time.

**Comment by tbnorth** [comment]
I don't think the style machinery survived very long. The original design had a tree based representation of the theme (colors, body elements, tree elements, etc.) which was compiled to a `@data qt-gui-plugin-style-sheet` by a @button script. I think it got broken when it was cleaned up and moved to the end of LeoSettings.leo.  So the source for the theme was the info. in the tree, and the `@data qt-gui-plugin-style-sheet` was just the compiled version.

Might be a bit laborious to try and bisect back to a good state, might be easier to try and fix the current pieces in place, bringing back any missing parts from git.  Haven't had any good chunks of Leo time recently.

**Comment by edreamleo** [comment]
@tbnorth Thanks for your comments. I think I agree with their general direction, but...

Haven't we both been forgetting that `@data qt-gui-plugin-style-sheet` is now "fully parameterized"?  This means that users don't need to change this stylesheet, they just need to change the various settings that it uses. In other words, we need to change the documentation.

Not at all sure about this.  What do you think, Terry?

**Comment by tbnorth** [comment]
Making arbitrary themes requires far more flexibility than can be achieved through parameterization in a single `@data qt-gui-plugin-style-sheet`.  Even if you were only changing colors, it would be hard to come up with a style-sheet that specified color for every single element that might need addressing to achieve a particular look.  The [+] widgets used to expand / contract, the appearance of radio buttons - they can all need tweaking.  My dark theme themes the next-month/last-month elements in calendar widgets. So themes must have their own style sheets.  Most flexibly, a theme should be able to operate at three levels:

 - just changes parameters
 - changes parameters and adds `@data qt-gui-theme-style-sheet`
 - changes parameters and replaces `@data qt-gui-plugin-style-sheet`

`@data qt-gui-theme-style-sheet` doesn't currently / has never existed, but would make sense when a theme wants to build on `@data qt-gui-plugin-style-sheet`.  There is currently `@data qt-gui-user-style-sheet`, but I think that should be left alone for local use by the user.

**Comment by edreamleo** [comment]
> Making arbitrary themes requires far more flexibility than can be achieved through parameterization in a single @data qt-gui-plugin-style-sheet.

@tbnorth It's been a long time since Leo included such style sheets.  I didn't find them in Leo 5.3 final(!) Do you know where they can be recovered?

**Comment by edreamleo** [comment]
@tbnorth I've just assigned you to this issue. I think I'll need your guidance.

I've also removed the "Blocker" label. We've lived with this issue a long time without questions or complaints from users.

**Comment by tbnorth** [comment]
On Sun, 05 Feb 2017 02:59:12 -0800
"Edward K. Ream" <notifications@github.com> wrote:

> @tbnorth I've just assigned you to this issue. I think I'll need your
> guidance.
> 
> I've also removed the "Blocker" label. We've lived with this issue a
> long time without questions or complaints from users.

There's been one or two reports, but fine to assign to me, will try and
get to it soon.

**Comment by edreamleo** [comment]
@tbnorth Thanks, Terry.

**Comment by LewisNeal** [comment]
@edreamleo Can I ask you to reconsider removing the "Blocker" label? The users (including me) have been very patient as many of the other areas of Leo improve and have bugs fixed. 
I think you might be surprised how many users would enjoy the themes/style problems being conquered.
Regards
Lewis

**Comment by edreamleo** [comment]
On Mon, Feb 20, 2017 at 8:21 PM, LewisNeal <notifications@github.com> wrote:

> @edreamleo <https://github.com/edreamleo> Can I ask you to reconsider
> removing the "Blocker" label?
>
​I agree that this item should be fixed asap. However, it's not a blocker
in the sense that I use the term.

Terry is on vacation at present, and I'll need to consult with him before
attempting a fix. I'll do this for 5.5 if possible, but no guarantees.

Edward

**Comment by edreamleo** [comment]
@LewisNeal Let me clarify that I do intend to fix this before Leo 5.5b1. You should worry only if I remove the 5.5 milestone from this item.

**Comment by edreamleo** [comment]
According to the [History of Leo](http://leoeditor.com/history.html), Leo 4.11 introduced themes in late 2013. I'll download the final versions of Leo since then and use the latest one that has the required `@button` node.  That will be the starting point, for reference.

**Comment by edreamleo** [comment]
Hmm.  It appears that all versions of Leo from 4.11 onward contain leo/config/themes.leo. I don't understand what, if anything, has changed from 4.11. Perhaps this is a documentation issue?

**Comment by edreamleo** [comment]
2e3916 completes this project.  [This post](https://groups.google.com/d/msg/leo-editor/cOmL8VwvOD8/RImJGuinAAAJ) discusses the changes made.

**Comment by LewisNeal** [comment]
See my comments at [#350: themes are ready for serious testing](https://groups.google.com/forum/?fromgroups#!topic/leo-editor/cOmL8VwvOD8)

**Comment by edreamleo** [comment]
I am going to reopen this. The reported problems are more than serious enough.

**Comment by edreamleo** [comment]
0a83517 adds a mind-boggling workaround. Urls must contain slashes, not back slashes, even on Windows.  Now expand/contract icons show up on Windows.  Presumably they never did before.

The stylesheet machinery is on the verge of being impossible to understand.  This is not entirely Leo's fault. Qt gives *no* hint about where syntax errors lie.

Settings substitutions within stylesheets can be ferocious to get right. I am not going to do anything about this now, but I am far from happy with the present morass.

Imo, Leo needs to dispense with most (all?) settings in style sheets.  This does *not* include syntax-coloring settings, which are entirely different.  Yes, this will make it harder to make systematic changes to, say, fonts.  But the present situation is intolerable.

After a long day's work I am finally able to work on individual themes.  That will take at least several more hours.

**Comment by tbnorth** [comment]
I would resist removing @setting expansion from stylesheets.  It's a major pain maintaining CSS without such a mechanism.  CSS3(?, maybe 2) has variables, but that doesn't Qt-CSS.

If the @setting expansion is hard to understand, let's blame docs., comments, and code layout, not the underlying mechanism. Struggling to understand an morass is sort of the way it is for code you didn't personally write recently, I know I feel that way quite often - let's not throw the baby out with the bathwater.

Cheers -Terry

**Comment by edreamleo** [comment]
The key to leaving this as they were (in code) was the fact that it is now possible to "build" `@data` nodes from their subtrees.  So there is no need for `@button make-theme`.  Instead, I added comments by hand in the subtrees with the following abbreviation:

    /*;;=/* ===== <|title|> ===== */

I also organized all `@theme` trees the same way, and "beautified" the various css settings.

Imo, the last two days work were exceptionally worthwhile.
